test/test_foreach.py

```diff
@@ -353,34 +353,60 @@ def test_int_scalar(self, device, dtype):
+                        continue
+
+                    res = foreach_bin_op(tensors, scalar)
+                    expected = [torch_bin_op(t, scalar) for t in tensors]
                     self.assertEqual(res, expected)
 
+                    # Test In-place
                     with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
                         foreach_bin_op_(tensors, scalar)
-                    return
 
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.div_(scalar) for t in tensors]
 
-                if foreach_bin_op_ == torch._foreach_div_ and dtype in torch.testing.integral_types() and self.device_type == "cpu":
-                    with self.assertRaisesRegex(RuntimeError,
-                                                "can't be cast to the desired output type"):
-                        foreach_bin_op_(tensors, scalar)
-                    return
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.mul_(scalar) for t in tensors]
 
-                # TODO[type promotion]: Fix once type promotion is enabled.
-                if dtype in torch.testing.integral_types() and self.device_type == 'cuda':
-                    self.assertEqual(res, [e.to(dtype) for e in expected])
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.add_(scalar) for t in tensors]
 
-                    foreach_bin_op_(tensors, scalar)
-                    self.assertEqual(tensors, [e.to(dtype) for e in expected])
-                else:
-                    self.assertEqual(res, expected)
-                    foreach_bin_op_(tensors, scalar)
-                    self.assertEqual(tensors, expected)
+                    with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator, with a bool"):
+                        [t.sub_(scalar) for t in tensors]
+                    continue
+
+                expected = [torch_bin_op(t, scalar) for t in tensors]
+                res = foreach_bin_op(tensors, scalar)
+
+                # In case of In-place division with integers, we can't change the dtype
+                if foreach_bin_op_ == torch._foreach_div_ and dtype in torch.testing.integral_types():
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        [t.div_(scalar) for t in tensors]
+
+                    with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                        torch._foreach_div_(tensors, scalar)
+                    continue
+
+                self.assertEqual(res, expected)
+
+                # In case of In-place op, we can't change the dtype
+                foreach_bin_op_(tensors, scalar)
```

Comment:
NB: this modifies the input "tensors" in-place, which means that the values are different the next iteration of the outer for-loop. But that seems to be okay. Usually to avoid mutating we would clone the input before performing the in-place operation.

No action item required here. It may possibly be nicer if we cloned `tensors` before calling the in-place op but that's subjective. I also see this pattern in the other test that is being modified in this PR.